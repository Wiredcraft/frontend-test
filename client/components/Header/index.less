.header {
  position: fixed;
  z-index: 1000;
  top: 0;
  width: 100%;
  height: 71px;
  background: #FFFFFF;
  box-shadow: 0px 4px 8px rgba(103, 102, 102, 0.25);

  .icons {
    height: 24px;
    position: absolute;
    right: 45px;
    width: 140px;
    top: 30px;
  
    .icon {
      &.home {
        margin-right: 35px;
      }
      &.ring {
        margin-right: 35px;
      }
    }
  }
}