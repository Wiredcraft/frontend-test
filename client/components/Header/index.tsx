import React from 'react';
import './index.less';

interface IProps {
  children: JSX.Element[]
}

export default function Header ({ children }: IProps): React.ReactElement {
  return (
    <div className="header" aria-label="header">
      {children}
    </div>
  )
};
