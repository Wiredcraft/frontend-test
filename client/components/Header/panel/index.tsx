import React from 'react';
import HomeIcon from '../../icons/homeIcon';
import RingIcon from '../../icons/ringIcon';
import AvatarIcon from '../../icons/avatarIcon';

export default function Panel(): React.ReactElement{
  return (
    <div className="icons">
      <span className="icon home"><HomeIcon /></span>
      <span className="icon ring"><RingIcon /></span>
      <span className="icon avatar"><AvatarIcon /></span>
    </div>
  );
}
 