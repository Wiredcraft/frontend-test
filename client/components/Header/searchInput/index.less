.search {
  position: absolute;
  width: 196px;
  height: 36px;
  right: 218px;
  top: 24px;
  background: #EAEAEA;
  border-radius: 19.5px;

  .icon {
    position: absolute;
    left: 11px;
    top: 8px;
  }

  .input {
    position: relative;
    width: 100%;
    height: 100%;

    input {
      width: 80%;
      height: 90%;
      display: block;
      border: none;
      border-radius: 19.5px;
      left: 20%;
      position: relative;
      background: #EAEAEA;

      &:focus {
        outline: none;
      }
    }
  }
}