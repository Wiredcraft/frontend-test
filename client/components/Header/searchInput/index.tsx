import React, { useState, useEffect } from 'react';
import { filterXss, debounce, getSearchParam } from 'utils';
import SearchIcon from '../../icons/searchIcon';
import { history } from '../../../index';
import './index.less';

export default function SearchInput(): React.ReactElement {
  const [search, setSearch] = useState(getSearchParam(history.location.search));

  useEffect(() => {
    history.listen(() => {
      const searchParam = getSearchParam(history.location.search);
      setSearch(searchParam);
    });
  }, [history.location.search]);

  const onInput = debounce((ev: React.ChangeEvent<HTMLInputElement>) => {
    const val = filterXss(ev.target.value);
    if (val) {
      history.push(`?search=${val}`)
    } else {
      history.push('')
    }
  }, 500);

  return (
    <div className="search">
      <div className="input">
        <span className="icon">
          <SearchIcon />
        </span>
        <input defaultValue={search} onInput={onInput} maxLength={20} aria-label="searchInput" />
      </div>
    </div>
  )
}