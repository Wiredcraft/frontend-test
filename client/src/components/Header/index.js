import React, {useState, useMemo} from 'react';
import {Link} from 'react-router-dom';
import {iconsMap} from '../../common/icons'

export default function Header(props) {

  const [menuClassName, setMenuClassName] = useState('');
  const [itemsClassName, setItemsClassName] = useState('');

  const toggleNav = (e) => {
    // show or hide navbar items
    setMenuClassName(menuClassName ? '' : 'show-items');
    // do animation
    setItemsClassName(itemsClassName ? '' : 'scale-fade-animation');
    // show or hide scroll bar
    document.body.style.overflowY = document.body.style.overflowY === 'hidden' ? 'scroll' : 'hidden';
  };

  const items = useMemo(() =>
      props.items.map((item, i) => {
        let html = item.html;
        if (html) {
          return html
        } else if (item.type === 'input') {
          html =
            <div className="search-input">
              {iconsMap[item.icon]}
              <input type="text" onChange={item.onChange}/>
            </div>
        } else if (item.type === 'link') {
          html = <Link to={item.to}>{item.text || iconsMap[item.icon]}</Link>
        }
        return (
          <li key={i} className={itemsClassName}>{html}</li>
        );
      }),
    [props.items, menuClassName, itemsClassName]);

  return (
    <div className="navbar">
      <h1></h1>

      <ul className={'items ' + menuClassName}>
        {items}
      </ul>

      <div className={'burger ' + menuClassName} onClick={toggleNav}>
        <div className="line1"/>
        <div className="line2"/>
        <div className="line3"/>
      </div>
    </div>
  )
}