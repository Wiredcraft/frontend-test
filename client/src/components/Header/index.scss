@import "../../sass/_variables";

.navbar {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 1rem 2rem .6rem;
  box-shadow: 5px 5px 5px $color-background;
  margin-bottom: 1rem;

  ul {
    display: flex;
    list-style: none;

    li {
      margin-left: 2em;

      a {
        text-decoration: none;
        color: $color-foreground;
        font-size: 1.5rem;
        line-height: $text-line-height;
      }
    }

    .search-input {
      background: $color-background;
      padding: 0 .5rem;
      border-radius: 1rem;
      font-size: 1rem;
      line-height: $text-line-height;
      color: $color-foreground;

      svg {
        vertical-align: middle;
        margin-right: 5px;
      }

      input {
        border: none;
        outline: none;
        background: transparent;
        line-height: $text-line-height;
        width: 7rem;
      }
    }
  }

  .burger {
    display: none;
    z-index: 1;
    cursor: pointer;

    div {
      background: black;
      height: 2px;
      width: 20px;
      margin-bottom: 3px;
      transition: 0.3s ease-in-out;
    }
  }

  @media (max-width: $mobile-max-width-md) {
    .burger {
      display: block;
    }

    .items {
      display: none;
      flex-direction: column;
      align-items: center;
      justify-content: space-around;
      background: white;
      height: 100%;
      width: 100%;
      position: absolute;
      top: 0;
      left: 0;
      transform: translateX(100%);
      transition: 0.3s ease-in-out;
      z-index: 1;

      li {
        margin-left: 0;
        transform: scale(0);
        opacity: 0;
      }
    }
  }
}

.show-items {
  transform: translateX(0) !important;
}

.show-items:not(div) {
  display: flex !important;
}

.show-items .line1 {
  transform: rotate(-45deg) translate(-3px, 5px);
}

.show-items .line2 {
  opacity: 0;
}

.show-items .line3 {
  transform: rotate(45deg) translate(-3px, -5px);
}

@keyframes scaleFade {
  from {
    transform: scale(0);
    opacity: 0;
  }

  to {
    transform: scale(1);
    opacity: 1;
  }
}

.scale-fade-animation {
  animation: scaleFade 0.3s ease-in-out forwards;
  animation-delay: 0.2s;
}