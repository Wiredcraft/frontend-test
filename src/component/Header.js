import React from 'react'
import avatarPic from '../asset/image/1/avatar.png';

const Header = () => {
  return (
    <header className="clearfix">
      <span className="avatar">
        <img src={avatarPic} alt="avatar" />
      </span>
      Reports
      <a>Specific</a>
      <a className="active">Overall</a>
    </header>
  )
}

export default Header