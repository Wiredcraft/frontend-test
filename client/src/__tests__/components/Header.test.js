import React from "react";
import {render, unmountComponentAtNode} from "react-dom";
import {act} from "react-dom/test-utils";
import {BrowserRouter} from "react-router-dom";
import Header from "../../components/Header";

let container = null;
beforeEach(() => {
  // create a DOM element as rendering target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // clear when unmount
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it("renders header", () => {
  const items = [
    {
      type: "input",
      icon: "search",
      onChange: (e) => {
      },
    },
    {
      type: "link",
      to: "/home",
      icon: "home",
    },
    {
      type: "link",
      to: "/notification",
      icon: "notification",
    },
    {
      type: "link",
      to: "/user",
      icon: "user",
    },
  ];

  act(() => {
    render(
      <BrowserRouter>
        <Header items={items}/>
      </BrowserRouter>,
      container
    );
  });

  expect(container.innerHTML).toMatchSnapshot();
});
