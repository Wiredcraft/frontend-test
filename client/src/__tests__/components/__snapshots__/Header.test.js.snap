// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders header 1`] = `"<div class=\\"navbar\\"><h1></h1><ul class=\\"items \\"><li class=\\"\\"><div class=\\"search-input\\"><svg stroke=\\"currentColor\\" fill=\\"none\\" stroke-width=\\"2\\" viewBox=\\"0 0 24 24\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><circle cx=\\"11\\" cy=\\"11\\" r=\\"8\\"></circle><line x1=\\"21\\" y1=\\"21\\" x2=\\"16.65\\" y2=\\"16.65\\"></line></svg><input type=\\"text\\"></div></li><li class=\\"\\"><a href=\\"/home\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 1024 1024\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M946.5 505L534.6 93.4a31.93 31.93 0 0 0-45.2 0L77.5 505c-12 12-18.8 28.3-18.8 45.3 0 35.3 28.7 64 64 64h43.4V908c0 17.7 14.3 32 32 32H448V716h112v224h265.9c17.7 0 32-14.3 32-32V614.3h43.4c17 0 33.3-6.7 45.3-18.8 24.9-25 24.9-65.5-.1-90.5z\\"></path></svg></a></li><li class=\\"\\"><a href=\\"/notification\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 24 24\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M12 22c1.1 0 2-.9 2-2h-4c0 1.1.89 2 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z\\"></path></svg></a></li><li class=\\"\\"><a href=\\"/user\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 496 512\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 96c48.6 0 88 39.4 88 88s-39.4 88-88 88-88-39.4-88-88 39.4-88 88-88zm0 344c-58.7 0-111.3-26.6-146.5-68.2 18.8-35.4 55.6-59.8 98.5-59.8 2.4 0 4.8.4 7.1 1.1 13 4.2 26.6 6.9 40.9 6.9 14.3 0 28-2.7 40.9-6.9 2.3-.7 4.7-1.1 7.1-1.1 42.9 0 79.7 24.4 98.5 59.8C359.3 421.4 306.7 448 248 448z\\"></path></svg></a></li></ul><div class=\\"burger \\"><div class=\\"line1\\"></div><div class=\\"line2\\"></div><div class=\\"line3\\"></div></div></div>"`;

exports[`should render header 1`] = `"<div class=\\"navbar\\"><h1></h1><ul class=\\"items \\"><li class=\\"\\"><div class=\\"search-input\\"><svg stroke=\\"currentColor\\" fill=\\"none\\" stroke-width=\\"2\\" viewBox=\\"0 0 24 24\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><circle cx=\\"11\\" cy=\\"11\\" r=\\"8\\"></circle><line x1=\\"21\\" y1=\\"21\\" x2=\\"16.65\\" y2=\\"16.65\\"></line></svg><input type=\\"text\\"></div></li><li class=\\"\\"><a href=\\"/home\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 1024 1024\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M946.5 505L534.6 93.4a31.93 31.93 0 0 0-45.2 0L77.5 505c-12 12-18.8 28.3-18.8 45.3 0 35.3 28.7 64 64 64h43.4V908c0 17.7 14.3 32 32 32H448V716h112v224h265.9c17.7 0 32-14.3 32-32V614.3h43.4c17 0 33.3-6.7 45.3-18.8 24.9-25 24.9-65.5-.1-90.5z\\"></path></svg></a></li><li class=\\"\\"><a href=\\"/notification\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 24 24\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M12 22c1.1 0 2-.9 2-2h-4c0 1.1.89 2 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z\\"></path></svg></a></li><li class=\\"\\"><a href=\\"/user\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 496 512\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 96c48.6 0 88 39.4 88 88s-39.4 88-88 88-88-39.4-88-88 39.4-88 88-88zm0 344c-58.7 0-111.3-26.6-146.5-68.2 18.8-35.4 55.6-59.8 98.5-59.8 2.4 0 4.8.4 7.1 1.1 13 4.2 26.6 6.9 40.9 6.9 14.3 0 28-2.7 40.9-6.9 2.3-.7 4.7-1.1 7.1-1.1 42.9 0 79.7 24.4 98.5 59.8C359.3 421.4 306.7 448 248 448z\\"></path></svg></a></li></ul><div class=\\"burger \\"><div class=\\"line1\\"></div><div class=\\"line2\\"></div><div class=\\"line3\\"></div></div></div>"`;
